Plugin fails when there's a name collision between a message and a service name.

`gnostic-grpc` exits with the following error:

```
$ gnostic foo.yaml --grpc-out=.
panic: interface conversion: desc.Descriptor is *desc.ServiceDescriptor, not *desc.MessageDescriptor

goroutine 1 [running]:
github.com/jhump/protoreflect/desc.(*FieldDescriptor).resolve(0xc000173980, 0xc0000c3860, 0x4, 0x8, 0xc00015f7d0, 0x2, 0x2, 0x2, 0x0)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/descriptor.go:552 +0x749
github.com/jhump/protoreflect/desc.(*MessageDescriptor).resolve(0xc00017a2d0, 0xc0000c3860, 0x3, 0x8, 0xc00015f7d0, 0x2, 0x2, 0x0, 0x0)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/descriptor.go:326 +0x498
github.com/jhump/protoreflect/desc.createFileDescriptor(0xc000134000, 0xc00000f800, 0x3, 0x3, 0x0, 0x2, 0xc0000c3b88, 0xc000116cf0)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/convert.go:88 +0x100c
github.com/jhump/protoreflect/desc.createFromSet(0xc000022d30, 0x9, 0x0, 0xc00015f750, 0x1, 0x1, 0xc0000c3b88, 0xc000116cf0, 0xc0000f8380, 0x0, ...)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/convert.go:225 +0x338
github.com/jhump/protoreflect/desc.createFileDescriptors(0xc00000f0c0, 0x4, 0x4, 0x0, 0x0, 0x0, 0x0)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/convert.go:131 +0x1d3
github.com/jhump/protoreflect/desc.createFileDescriptorsFromSet(0xc000032680, 0x0, 0x0, 0x0, 0x0)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/convert.go:193 +0xa7
github.com/jhump/protoreflect/desc.createFileDescriptorFromSet(0xc000032680, 0x0, 0x0, 0x0, 0x0)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/convert.go:169 +0x39
github.com/jhump/protoreflect/desc.CreateFileDescriptorFromSet(...)
        /usr/local/google/home/timburks/go/src/github.com/jhump/protoreflect/desc/convert.go:165
github.com/googleapis/gnostic-grpc/generator.(*Renderer).RenderProto(0xc000032640, 0xc000032680, 0xc000022d20, 0x9, 0xc000022d20, 0xc000022d23, 0xc0000c3e98)
        /usr/local/google/home/timburks/go/src/github.com/googleapis/gnostic-grpc/generator/renderer.go:82 +0x49
github.com/googleapis/gnostic-grpc/generator.(*Renderer).Render(0xc000032640, 0xc000106bd0, 0xc000022d20, 0x9, 0x6, 0xc000022d20)
        /usr/local/google/home/timburks/go/src/github.com/googleapis/gnostic-grpc/generator/renderer.go:62 +0x96
github.com/googleapis/gnostic-grpc/generator.RunProtoGenerator()
        /usr/local/google/home/timburks/go/src/github.com/googleapis/gnostic-grpc/generator/main.go:66 +0x47c
main.main()
        /usr/local/google/home/timburks/go/src/github.com/googleapis/gnostic-grpc/main.go:22 +0x20
Errors reading foo.yaml
exit status 2
```
for the following `foo.yaml` input file:
```
# Specification for the Foo Service API
swagger: '2.0'
info:
  title: Foo Service API
  description: Foo Service API
  version: '0.0.1'
host: api.foo.com
schemes:
  - https
  - http
produces:
  - application/json
basePath: /
paths:

  /{foo-id}:
    parameters:
      - in: path
        name: foo-id
        type: string
        required: true
    get:
      tags:
        - Foo
      description: Retrieve metadata about the foo.
      responses:
        '200':
          description: Description of the foo
          schema:
            $ref: '#/definitions/Foo'

    delete:
      tags:
        - Foo
      description: Remove the foo.
      responses:
        '200':
          description: Description of the foo
          schema:
            $ref: '#/definitions/Foo'

definitions:
  Foo:
    type: object
    required:
      - name
    properties:
      self:
        description: A link to this foo.
        type: string
        format: uri
      kind:
        description: Should always be "Foo".
        type: string
      name:
        description: Name of the foo. Required in order to create the foo.
        type: string
      created_at:
        description: Timestamp in milliseconds since the Unix Epoch.
        type: integer
        format: int64
      lastmodified_at:
        description: Timestamp in milliseconds since the Unix Epoch.
        type: integer
        format: int64
```

Notably, if `foo.yaml` is renamed to `bar.yaml`, the plugin succeeds.

From the error message and the generated `.proto` files, I suspect that the problem is related to a message and service both having the name `Foo` (gnostic-grpc derives the service name from the input file name).

--- a/generator/generator.go
+++ b/generator/generator.go
@@ -241,7 +241,7 @@ func buildMessagesFromTypes(descr *dpb.FileDescriptorProto, renderer *Renderer) 
 // have to be set.
 func buildServiceFromMethods(descr *dpb.FileDescriptorProto, renderer *Renderer) (err error) {
 	methods := renderer.Model.Methods
-	serviceName := strings.Title(renderer.Package)
+	serviceName := findValidServiceName(descr.MessageType, strings.Title(renderer.Package))
 
 	service := &dpb.ServiceDescriptorProto{
 		Name: &serviceName,
@@ -648,4 +648,27 @@ func convertStatusCodes(name string) string {
 		name = strings.Replace(statusText, " ", "_", -1)
 	}
 	return name
+}
+
+// Finds a valid service name for the gRPC service. A valid service name is not already taken by a message
+// Reference: https://github.com/googleapis/gnostic-grpc/issues/7
+func findValidServiceName(messages []*dpb.DescriptorProto, serviceName string) string {
+	messageNames := make(map[string]bool)
+
+	for _, m := range messages {
+		messageNames[*m.Name] = true
+	}
+
+	validServiceName := serviceName
+	ctr := 0
+	for {
+		if nameIsAlreadyTaken, _ := messageNames[validServiceName]; !nameIsAlreadyTaken {
+			return validServiceName
+		}
+		validServiceName = serviceName + "Service"
+		if ctr > 0 {
+			validServiceName += strconv.Itoa(ctr)
+		}
+		ctr += 1
+	}
 }

--- a/generator/generator.go
+++ b/generator/generator.go
@@ -41,6 +41,8 @@ var generatedSymbolicReferences = make(map[string]bool, 0)
 // Gathers all messages that have been generated from symbolic references in recursive calls.
 var generatedMessages = make(map[string]string, 0)
 
+var shouldRenderEmptyImport = false
+
 // Uses the output of gnostic to return a dpb.FileDescriptorSet (in bytes). 'renderer' contains
 // the 'model' (surface model) which has all the relevant data to create the dpb.FileDescriptorSet.
 // There are four main steps:
@@ -70,8 +72,6 @@ func (renderer *Renderer) runFileDescriptorSetGenerator() (fdSet *dpb.FileDescri
 		return nil, err
 	}
 
-	addDependencies(fdSet)
-
 	err = buildMessagesFromTypes(mainProto, renderer)
 	if err != nil {
 		return nil, err
@@ -82,16 +82,24 @@ func (renderer *Renderer) runFileDescriptorSetGenerator() (fdSet *dpb.FileDescri
 		return nil, err
 	}
 
+	addDependencies(fdSet)
+
 	return fdSet, err
 }
 
-// Adds the dependencies to the FileDescriptor we want to render. This essentially makes the 'import' statements
-// inside the .proto definition.
+// Adds the dependencies to the FileDescriptorProto we want to render (the last one). This essentially makes the 'import'
+// statements inside the .proto definition.
 func addDependencies(fdSet *dpb.FileDescriptorSet) {
 	// At last, we need to add the dependencies to the FileDescriptorProto in order to get them rendered.
 	lastFdProto := getLast(fdSet.File)
 	for _, fd := range fdSet.File {
 		if fd != lastFdProto {
+			if *fd.Name == "google/protobuf/empty.proto" { // Reference: https://github.com/googleapis/gnostic-grpc/issues/8
+				if shouldRenderEmptyImport {
+					lastFdProto.Dependency = append(lastFdProto.Dependency, *fd.Name)
+				}
+				continue
+			}
 			lastFdProto.Dependency = append(lastFdProto.Dependency, *fd.Name)
 		}
 	}
@@ -264,9 +272,11 @@ func buildServiceFromMethods(descr *dpb.FileDescriptorProto, renderer *Renderer)
 
 		if method.ParametersTypeName == "" {
 			method.ParametersTypeName = "google.protobuf.Empty"
+			shouldRenderEmptyImport = true
 		}
 		if method.ResponsesTypeName == "" {
 			method.ResponsesTypeName = "google.protobuf.Empty"
+			shouldRenderEmptyImport = true
 		}
 
 		mDescr := &dpb.MethodDescriptorProto{

--- a/generator/generator.go
+++ b/generator/generator.go
@@ -222,7 +222,7 @@ func buildMessagesFromTypes(descr *dpb.FileDescriptorProto, renderer *Renderer) 
 			// Maps are represented as nested types inside of the descriptor.
 			if f.Kind == surface_v1.FieldKind_MAP {
 				if strings.Contains(f.Type, "map[string][]") {
-					// Not supported for now: https://github.com/LorenzHW/gnostic-grpc/issues/3#issuecomment-509348357
+					// Not supported for now: https://github.com/LorenzHW/gnostic-grpc-deprecated/issues/3#issuecomment-509348357
 					continue
 				}
 				mapDescriptorProto := buildMapDescriptorProto(f)

